### PR TITLE
Forward Mode CallInst known functions

### PR DESCRIPTION
**Pull request by tgymnich**
- [x] known functions
- [x] tests

### CODE REVIEW

File: enzyme/Enzyme/AdjointGenerator.h [hunk 1]
```diff
@@ -5821,31 +5940,54 @@ class AdjointGenerator
...
       }
 
       // Functions that only modify pointers and don't allocate memory,
       // needs to be run on shadow in primal
       if (funcName == "_ZSt29_Rb_tree_insert_and_rebalancebPSt18_Rb_tree_"
                       "node_baseS0_RS_") {
-        if (Mode == DerivativeMode::ReverseModeGradient) {
+        if (Mode == DerivativeMode::ReverseModeGradient ||
+            Mode == DerivativeMode::ForwardMode) {
           eraseIfUnused(*orig, /*erase*/ true, /*check*/ false);
```
Comment (wsmoses) [inline enzyme/Enzyme/AdjointGenerator.h:5991]: This still needs to be run in forward mode, whose implementation should be essentially the same as the augmented primal.

File: enzyme/Enzyme/AdjointGenerator.h [hunk 2]
```diff
@@ -5954,29 +6096,56 @@ class AdjointGenerator
             }
```
Comment (wsmoses) [inline enzyme/Enzyme/AdjointGenerator.h:6096]: The dynamic_cast code above needs modification for forward mode as well.

File: enzyme/Enzyme/AdjointGenerator.h [hunk 3]
```diff
@@ -6131,39 +6411,72 @@ class AdjointGenerator
...
+            return;
+          }
+          }
         }
 
         if (funcName == "julia.write_barrier") {
-          if (Mode == DerivativeMode::ReverseModeGradient) {
+          if (Mode == DerivativeMode::ReverseModeGradient ||
+              Mode == DerivativeMode::ForwardMode) {
             eraseIfUnused(*orig, /*erase*/ true, /*check*/ false);
```
Comment (wsmoses) [inline enzyme/Enzyme/AdjointGenerator.h:6480]: The write barrier needs to also be run on shadow pointers on the forward (as below)

File: enzyme/Enzyme/AdjointGenerator.h [hunk 4]
```diff
@@ -6388,7 +6790,8 @@ class AdjointGenerator
       // TODO enable this if we need to free the memory
       // NOTE THAT TOPLEVEL IS THERE SIMPLY BECAUSE THAT WAS PREVIOUS ATTITUTE
       // TO FREE'ing
-      if (Mode != DerivativeMode::ReverseModeCombined) {
+      if (Mode == DerivativeMode::ReverseModeGradient ||
```
Comment (wsmoses) [inline enzyme/Enzyme/AdjointGenerator.h:6793]: These changes modify the behavior of reverse mode for allocations. Moreover forward mode support should be added here, which should essentially keep the same mallocs and free as the original code, duplicating as required for shadows.

File: enzyme/Enzyme/AdjointGenerator.h [hunk 5]
```diff
@@ -5887,7 +6028,8 @@ class AdjointGenerator
           if (subretType == DIFFE_TYPE::DUP_ARG) {
             Value *shadow = placeholder;
             if (lrc || Mode == DerivativeMode::ReverseModePrimal ||
-                Mode == DerivativeMode::ReverseModeCombined) {
+                Mode == DerivativeMode::ReverseModeCombined ||
+                Mode == DerivativeMode::ForwardMode) {
```
Comment (wsmoses) [inline enzyme/Enzyme/AdjointGenerator.h:6032]: Also may need to mod 6051

File: enzyme/Enzyme/AdjointGenerator.h [hunk 6]
```diff
@@ -5925,6 +6067,12 @@ class AdjointGenerator
           }
         }
 
+        if (Mode == DerivativeMode::ForwardMode) {
+          eraseIfUnused(*orig);
+          assert(gutils->isConstantInstruction(orig));
```
Comment (wsmoses) [inline enzyme/Enzyme/AdjointGenerator.h:6072]: Add an unreachable and print error as not yet implemented.

File: enzyme/Enzyme/AdjointGenerator.h [hunk 7]
```diff
@@ -6425,7 +6846,7 @@ class AdjointGenerator
             freeKnownAllocation(Builder2, lookup(nop, Builder2), *called,
                                 dbgLoc, gutils->TLI);
           }
-        } else if (Mode != DerivativeMode::ReverseModePrimal) {
+        } else {
```
Comment (wsmoses) [inline enzyme/Enzyme/AdjointGenerator.h:6849]: This should be != primal and != forward

File: enzyme/Enzyme/AdjointGenerator.h [hunk 8]
```diff
@@ -6549,7 +6970,7 @@ class AdjointGenerator
       // TO FREE'ing
       if (Mode == DerivativeMode::ReverseModeGradient) {
         eraseIfUnused(*orig, /*erase*/ true, /*check*/ false);
-      } else if (Mode != DerivativeMode::ReverseModeCombined) {
+      } else if (Mode == DerivativeMode::ReverseModePrimal) {
```
Comment (wsmoses) [inline enzyme/Enzyme/AdjointGenerator.h:6973]: or == Gradient

File: enzyme/Enzyme/AdjointGenerator.h [hunk 9]
```diff
@@ -6584,6 +7005,16 @@ class AdjointGenerator
       assert(gutils->invertedPointers.find(orig) ==
              gutils->invertedPointers.end());
 
+      if (Mode == DerivativeMode::ForwardMode) {
+        IRBuilder<> Builder2(&call);
```
Comment (wsmoses) [inline enzyme/Enzyme/AdjointGenerator.h:7009]: You need to check if !isConstantValue(orig->getArgOperand(0))

and also invertPointer

Comment (wsmoses) [reply enzyme/Enzyme/AdjointGenerator.h:7009]: Add a runtime check that inverted ptr != original ptr

**Review by wsmoses — APPROVED** [review]

File: enzyme/Enzyme/AdjointGenerator.h (hunk 7 above)
Comment (tgymnich) [reply enzyme/Enzyme/AdjointGenerator.h:6849]: done 5258db043a9b229393ca3d2e2f79232a2bc32b62

File: enzyme/Enzyme/AdjointGenerator.h (hunk 9 above)
Comment (tgymnich) [reply enzyme/Enzyme/AdjointGenerator.h:7010]: done 3a04ef2142b44e66876b68c2b559b660f7929533

**Pull request merged**